ls1intum/Artemis
# Feature/programming exercise/simplify submission and result generation in the local setup

The following feature is only intended for local usage, in particular when no local version control and continuous integration is setup.




### Checklist

### Motivation and Context


The creation of a programming exercise in the development environment requires a local setup of a VCS and a CI. This setup needs a lot of time and not all developers need the full functionality of the programming exercises. To solve this issue this PR enables the creation of programming exercises and the simulation of participation's, submissions and results without a local setup. The introduced changes can only be used in the dev environment.
### Description

1. Added a checkbox "no local setup" inside of the creation form of the programming exercises, which is only visible in the test and dev environment. When the user checks this box and clicks on generate the programming exercise is generated without the connection to the VCS and CI. 
2. Added a new Resource "ProgrammingSubmissionResultSimulationResource" which simulates participations, submissions and results for the current logged in user. The simulation is triggered by the click on the newly introduced buttons "submission" and "result". These buttons are also only visible in the dev and test environment.
### Steps for Testing

This feature can only be tested in your local dev environment (i.e. the `dev` profile needs to be active, otherwise the new buttons and REST Calls won't be available)

0. Checkout the branch and start Artemis in IntelliJ
1. Navigate to Course Administration
(1.1 Create a Course if you don't have one) 
2. Click on the Exercise button of a course
3. Click on Generate new programming exercise button
4. Insert the required fields
5. Check the Checkbox "No Local Setup"
6. Click on the Generate button 

-> now the programming exercise is created, check the database tables(exercise, participation, submission and result) if the new exercise as well as the base and solution participation, submission and result are created 
![ExerciseSimulation](https://user-images.githubusercontent.com/12066497/78762108-c02be400-7983-11ea-9d5e-73fa7051f83c.gif)

7. Navigate to the course overview
8. Click on the course for which you created the programming exercise
9. Click on the submission button

-> now a new participation and a submission are created, check the tables(participation and submission) if the new participation and submission are created
![SubmissionSimulation](https://user-images.githubusercontent.com/12066497/78762937-d6866f80-7984-11ea-98c3-811cea4a2392.gif)

10. Click on the result button

-> now a new result is created, check the result table if the new result is created 
![ResultSubmission](https://user-images.githubusercontent.com/12066497/78762955-db4b2380-7984-11ea-89bb-9e8d2f5d7949.gif)

### Screenshots


1. The newly introduced checkbox. If you tick this check box and click on the generate button the new exercise will be created without a real VCS and CI connection
![noLocalSetupCheckbox](https://user-images.githubusercontent.com/12066497/78711455-4bbb5b80-7917-11ea-8c51-b2ed38e4052d.png)
2. The new submission and result simulation buttons. If you click on submission button, a new participation and submission will be simulated and stored in the database. If you click on the result button, a new result will be generated based on the last stored simulated submission
![NewButtons](https://user-images.githubusercontent.com/12066497/78711502-5b3aa480-7917-11ea-85d4-5836b53e070a.png)
